FlowCrypt/Functionality/Pgp/KeyMethods.swift

```diff
@@ -63,8 +41,45 @@ final class KeyMethods: KeyMethodsType {
             armoredOrBinary: armored.joined(separator: "\n").data()
         )
         guard parsed.keyDetails.count == armored.count else {
-            throw KeyServiceError.parsingError
+            throw KeypairError.parsingError
         }
         return parsed.keyDetails
     }
+
+    func chooseSenderSigningKey(keys: [Keypair], senderEmail: String) async throws -> Keypair? {
+        let parsed = try await parseKeys(armored: keys.map { $0.private })
+        guard let usable = parsed.first(where: { $0.isKeyUsable }) else { return nil }
+        return keys.first { $0.primaryFingerprint == usable.primaryFingerprint }
+    }
+
+    func chooseSenderEncryptionKeys(keys: [Keypair], senderEmail: String) async throws -> [Keypair] {
+        let senderEmail = senderEmail.lowercased()
+        let parsed = try await parseKeys(armored: keys.map { $0.public })
+        guard parsed.isNotEmpty else {
+            throw KeypairError.noAccountKeysAvailable
+        }
+        let usable = parsed.filter { $0.isKeyUsable }
+        guard usable.isNotEmpty else {
+            throw MessageValidationError.noUsableAccountKeys
+        }
+        if let byPrimaryUid = filter(keys, usable, ({ $0.pgpUserEmailsLowercased.first == senderEmail })) {
+            return byPrimaryUid // if any keys match by primary uid, use them
+        }
+        if let byAnyUid = filter(keys, usable, ({ $0.pgpUserEmailsLowercased.contains(senderEmail) })) {
+            return byAnyUid // if any keys match by any uid, use them
+        }
+        return filter(keys, usable, { _ in true }) ?? [] // use any usable keys
+    }
+
+    private func filter(_ toReturn: [Keypair], _ parsed: [KeyDetails], _ criteria: (KeyDetails) -> Bool) -> [Keypair]? {
+        let filteredPrimaryFingerprints = parsed
+            .filter { criteria($0) }
+            .map { $0.primaryFingerprint }
+        let filteredKeypairs = toReturn
+            .filter { filteredPrimaryFingerprints.contains($0.primaryFingerprint) }
+        guard filteredKeypairs.isNotEmpty else {
+            return nil
+        }
+        return filteredKeypairs
```

Comment:
Do we need to return nil when array is empty when return type is array?
I think we don't need above guard  and `return filteredKeypairs` is enough.
And we can change `if let byPrimaryUid = filter(keys, usable, ({ $0.pgpUserEmailsLowercased.first == senderEmail }))` to `if let byPrimaryUid = filter(keys, usable, ({ $0.pgpUserEmailsLowercased.first == senderEmail })), byPrimaryUid.isNotEmpty`